### ISSUE TITLE
CUDA OOM trying to LoRA a 7B model on distributed setting

### ISSUE THREAD

**Issue by lulmer** [pytorch/torchtune#2078]
Hello, 
My issue is essentially related to the one that has been solved yesterday by @joecummings on #2071. 
I have a dataset of around 70 000 samples in an openai chat format stored as an json file, after an analysis with a tokenizer I can tell that the sequence length of the inputs (the "text" column obtained after doing `apply_chat_template`) is spanning from 50 tokens to 19900 tokens.
I tried to copy the provided config for Qwen2.5-7B and modified it so it uses my dataset. When I try to launch my training on two H100 on the same node I have a CUDA OOM error. 
I have noticed that I can train end-to-end this dataset on a single GPU config only if I enable `activation_gradient_checkpointing` (uses arround 91-97% of the vRAM though) otherwise I run straight into OOM as well. 
With `activation_gradient_checkpointing` enabled on the multigpu the training will start and raise OOM after 15% of it. 

Here is the config I am using : 
```yaml
# Config for multi-device LoRA finetuning in lora_finetune_distributed.py
# using a Qwen2.5 7B model
#
# This config assumes that you've run the following command before launching
# this run:
#   tune download Qwen/Qwen2.5-7B-Instruct --output-dir /tmp/Qwen2_5-7B-Instruct
#
# To launch on 2 devices, run the following command from root:
#   tune run --nnodes 1 --nproc_per_node 2 lora_finetune_distributed --config qwen2_5/7B_lora
#
# You can add specific overrides through the command line. For example
# to override the checkpointer directory while launching training
# you can run:
#   tune run --nnodes 1 --nproc_per_node 2 lora_finetune_distributed --config qwen2_5/7B_lora checkpointer.checkpoint_dir=<YOUR_CHECKPOINT_DIR>
#
# This config works best when the model is being fine-tuned on 2+ GPUs.
# For single device LoRA finetuning please use 7B_lora_single_device.yaml



# Model Arguments
model:
  _component_: torchtune.models.qwen2_5.lora_qwen2_5_7b_instruct
  lora_attn_modules: ['q_proj', 'v_proj', 'output_proj']
  apply_lora_to_mlp: True
  apply_lora_to_output: False
  lora_rank: 8  # higher increases accuracy and memory
  lora_alpha: 16  # usually alpha=2*rank
  lora_dropout: 0.0

tokenizer:
  _component_: torchtune.models.qwen2_5.qwen2_5_tokenizer
  path: /data/checkpoints/Qwen-Qwen2.5-7B-Instruct/vocab.json
  merges_file: /data/checkpoints/Qwen-Qwen2.5-7B-Instruct/merges.txt
  max_seq_len: 20000

checkpointer:
  _component_: torchtune.training.FullModelHFCheckpointer
  checkpoint_dir: /data/checkpoints/Qwen-Qwen2.5-7B-Instruct/
  checkpoint_files: [
    model-00001-of-00004.safetensors,
    model-00002-of-00004.safetensors,
    model-00003-of-00004.safetensors,
    model-00004-of-00004.safetensors,
  ]
  recipe_checkpoint: null
  output_dir: /data/checkpoints/MyCustomModel-Qwen2.5-7B-Instruct/
  model_type: QWEN2
resume_from_checkpoint: False

# Dataset and Sampler
dataset:
  _component_: torchtune.datasets.chat_dataset
  source: json
  conversation_column: messages
  conversation_style: openai
  data_files: /data/datasets/my_custom_dataset.json
  split: train
  packed: False  # True increases speed
seed: null
shuffle: True
batch_size: 2

# Optimizer and Scheduler
optimizer:
  _component_: torch.optim.AdamW
  fused: True
  weight_decay: 0.01
  lr: 3e-4
lr_scheduler:
  _component_: torchtune.training.lr_schedulers.get_cosine_schedule_with_warmup
  num_warmup_steps: 100

loss:
  _component_: torchtune.modules.loss.CEWithChunkedOutputLoss

# Training
epochs: 3
max_steps_per_epoch: null
gradient_accumulation_steps: 8  # Use to increase virtual batch size
compile: False  # pytorch compile, set to true for better perf/memory

# Logging
output_dir: /data/checkpoints/trained_checkpoints/MyCustomModel-Qwen2.5-7B-Instruct
metric_logger:
  _component_: torchtune.training.metric_logging.TensorBoardLogger
  log_dir: ${output_dir}
log_every_n_steps: 1
log_peak_memory_stats: False

# Environment
device: cuda
dtype: bf16

# Activations Offloading
enable_activation_checkpointing: False  # True reduces memory
enable_activation_offloading: False  # True reduces memory

# Show case the usage of pytorch profiler
# Set enabled to False as it's only needed for debugging training
profiler:
  _component_: torchtune.training.setup_torch_profiler
  enabled: False

  #Output directory of trace artifacts
  output_dir: ${output_dir}/profiling_outputs

  #`torch.profiler.ProfilerActivity` types to trace
  cpu: True
  cuda: True

  #trace options passed to `torch.profiler.profile`
  profile_memory: False
  with_stack: False
  record_shapes: True
  with_flops: False

  # `torch.profiler.schedule` options:
  # wait_steps -> wait, warmup_steps -> warmup, active_steps -> active, num_cycles -> repeat
  wait_steps: 5
  warmup_steps: 5
  active_steps: 2
  num_cycles: 1
```
Here is the command I am using : 
```bash
CUDA_VISIBLE_DEVICES=6,7 tune run --nnodes 1 --nproc_per_node 2 lora_finetune_distributed --config user_configs/custom_model_qwen.yaml
```

Here is the error I am getting on multigpu setting : 
```bash
[rank0]: torch.OutOfMemoryError: CUDA out of memory. Tried to allocate 550.00 MiB. GPU 0 has a total capacity of 79.11 GiB of which 286.94 MiB is free. Process 4048723 has 78.82 GiB memory in use. Of the allocated memory 72.28 GiB is allocated by PyTorch, and 4.76 GiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)
1|3|Loss: 0.6639634966850281:   0%|                                                                                               | 3/2367 [00:07<1:41:18,  2.57s/it]
W1127 09:37:49.261000 1902624 torch/distributed/elastic/multiprocessing/api.py:897] Sending process 1902959 closing signal SIGTERM
E1127 09:37:49.627000 1902624 torch/distributed/elastic/multiprocessing/api.py:869] failed (exitcode: 1) local_rank: 0 (pid: 1902958) of binary: /workspaces/torchtune_exploration/.venv/bin/python
Traceback (most recent call last):
  File "/workspaces/torchtune_exploration/.venv/bin/tune", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/workspaces/torchtune_exploration/torchtune/torchtune/_cli/tune.py", line 49, in main
    parser.run(args)
  File "/workspaces/torchtune_exploration/torchtune/torchtune/_cli/tune.py", line 43, in run
    args.func(args)
  File "/workspaces/torchtune_exploration/torchtune/torchtune/_cli/run.py", line 212, in _run_cmd
    self._run_distributed(args, is_builtin=is_builtin)
  File "/workspaces/torchtune_exploration/.venv/lib/python3.11/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 355, in wrapper
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/workspaces/torchtune_exploration/torchtune/torchtune/_cli/run.py", line 101, in _run_distributed
    run(args)
  File "/workspaces/torchtune_exploration/.venv/lib/python3.11/site-packages/torch/distributed/run.py", line 910, in run
    elastic_launch(
  File "/workspaces/torchtune_exploration/.venv/lib/python3.11/site-packages/torch/distributed/launcher/api.py", line 138, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/torchtune_exploration/.venv/lib/python3.11/site-packages/torch/distributed/launcher/api.py", line 269, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError: 
============================================================
/workspaces/torchtune_exploration/torchtune/recipes/lora_finetune_distributed.py FAILED
```

To me it seems very odd because I have a lower memory footprint with `unsloth` or `trl` while finetuning on more layers with higher `r`. Am I doing something wrong ?

**Comment by felipemello1** [comment]
Hey @lulmer, if it’s OOMing in the middle of training, my main hypothesis is are getting hit with a huge sequence. It is not possible to fit 20k context length with 7b in 2 gpus without activation checkpointing, because the activations are too large.

A simple way to verify it is using dataset.packex=True, which will pack multiple sequences together in the same sample; making every sample have your tokenized.max_seq_len and maximizing you gpu utilization.

there are some low hanging fruits in your config. Please consider taking a look at our optimization flags https://github.com/pytorch/torchtune?tab=readme-ov-file#optimization-flags

more specifically, set dataset.packed=true, compile=true, and use the max_seq_len as a multiple of 64.

**Comment by felipemello1** [comment]
One last thing you can add is enabling “expandable segments = True” environment flag, which avoids memory fragmentation due to variable seq Len and should help with OOMing

**Comment by felipemello1** [comment]
Btw, don’t forget to tune your learning rate. Usually with Lora and/or higher batch size you can 10x it and have better loss

**Comment by felipemello1** [comment]
> Do you have any recommendations on how I can programmatically sweep through different learning rates and log all of that to select the best hyperparams ?

Not programatically. What I personally do sometimes is write the different commands in a .py file and run it with subprocess. Something like:

```
import subprocess
a = """tune run <rest of command>"""
b = ..
c = ...
commands = [a,b,d]
for command in commands:
	subprocess.run(command, shell=True)
```

To log, i recommend using weights and biases: https://pytorch.org/torchtune/main/generated/torchtune.training.metric_logging.WandBLogger.html?highlight=wandb#torchtune.training.metric_logging.WandBLogger

Then there it for N steps (eg. 200), and lr like 1e-3, 1e-4, 1e-5

>  I am familiar with the idea of Rotary Positional Embeddings but do you have a clue of the kind of black magic he is doing?

By magic, do you mean ROPE or not OOMing?

If you are referring to ROPE, you can check our implementation here: https://github.com/pytorch/torchtune/blob/160fd96857483b343b4440c0db36692cc5a83a3d/torchtune/models/llama3_1/_position_embeddings.py#L15

If you mean how to get to 128k: 
Memory has 4 main components: model, optimizer state, gradients, activations
If you finetune QLoRA, model is quantized, and optimizer state and gradients are small, because you are not finetuning much. The main issue becomes activations. With checkpointing and offloading, they reduce drastically. Aside from that, we also do chunked cross entropy to avoid the peak memory when computing the loss.

**Comment by ebsmothers** [comment]
@lulmer just to tack onto what @felipemello1 already said about RoPE scaling, I think there are two separate points. One is the ability to finetune a model with long context length. Here the activation memory becomes the bottleneck and so you can set `enable_activation_checkpointing=True` and `enable_activation_offloading=True` to finetune on a long context length without OOMing. 

Separately, RoPE scaling is a technique to finetune or run inference on a longer context length than the model was originally trained on. You can think of RoPE as applying some rotation to different tokens based on their relative position to one another in the sequence. This corresponds to spacing relative positions between any two tokens out along the circle and requires knowing the maximum sequence length of the model in advance. If we want to extend a model trained with e.g. max sequence length 2048 to a dataset with sequence length 4096, we can interpolate from 2048 points to 4096 along the circle by extending the original RoPE embeddings (and there are different ways this interpolation can be done, hence different methods of RoPE scaling). This is a very handwavy explanation, but would strongly recommend [this section](https://blog.eleuther.ai/yarn/#position-interpolation) (and the image contained within) from a blog by Eleuther on RoPE. And if you want to go deeper, there is a ton more detail throughout the rest of the blog too.

To summarize, RoPE scaling is not really the means by which you avoid OOMs in long-context finetuning, that would be activation checkpointing and offloading (among other techniques). But RoPE scaling is a modeling technique that allows you to transfer a pretrained model to a long-context task, even if the original model was not trained using long-context data.

**Comment by fabiogeraci** [comment]
I am trying to fine-tune the Llama3.1 8B using and adapted [Llama 3.1 70B distributed](https://github.com/pytorch/torchtune/blob/main/recipes/configs/llama3_1/70B_full.yaml) config and the [full_finetune_distributed](https://github.com/pytorch/torchtune/blob/main/recipes/full_finetune_distributed.py)

```
torch.distributed.elastic.multiprocessing.errors.ChildFailedError: 
============================================================
src.full_finetune_distributed FAILED
------------------------------------------------------------
Failures:
  <NO_OTHER_FAILURES>
------------------------------------------------------------
Root Cause (first observed failure):
[0]:
  time      : 2024-12-04_09:46:08
  host      : farm22-gpu0303.internal.sanger.ac.uk
  rank      : 0 (local_rank: 0)
  exitcode  : 1 (pid: 3889705)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
============================================================
```

[8B_full_distributed.txt](https://github.com/user-attachments/files/18005982/8B_full_distributed.txt)
[full_trace.txt](https://github.com/user-attachments/files/18005999/full_trace.txt)
